### DISCUSSION TITLE
AttributeError: module 'netket.hilbert' has no attribute 'constraint'

### DISCUSSION THREAD

**Discussion by alex-shilcusky** [netket/netket#1970] [Q&A] [answered]
I am trying to write my own Hilbert space class, which will be similar to the spin class. In this class, I want to use DiscreteHilbertConstraint and SumConstraint. I follow the example found here:
https://netket.readthedocs.io/en/latest/api/_generated/hilbert/netket.hilbert.constraint.DiscreteHilbertConstraint.html

but when doing so I get the error 
> AttributeError: module 'netket.hilbert' has no attribute 'constraint'

at the line 

> nk.hilbert.constraint.DiscreteHilbertConstraint

I am on Netket 3.15.1. I would appreciate any advice on how to use the constraint class in a locally-defined hilbert class. Thanks.

**Comment by PhilipVinc** [comment, marked as answer]
Can you share the whole error? 
And can you share the exact version of netket installed by doing `pip freeze | grep netket` or equivalent command?

If I run on the last (3.15.1) netket the following code runs fine:
```python
import netket as nk
from netket.utils import struct

import jax; import jax.numpy as jnp

class SumConstraint(nk.hilbert.constraint.DiscreteHilbertConstraint):
    # A simple constraint checking that the total sum of the elements
    # in the configuration is equal to a given value.

    # The value must be set as a pytree_node=False field, meaning
    # that it is a constant and changes to this value represent different
    # constraints.
    total_sum : float = struct.field(pytree_node=False)

    def __init__(self, total_sum):
        self.total_sum = total_sum

    def __call__(self, x):
        # Makes it jax-compatible
        return jnp.sum(x, axis=-1) == self.total_sum

    def __hash__(self):
        return hash(("SumConstraint", self.total_sum))

    def __eq__(self, other):
        if isinstance(other, SumConstraint):
            return self.total_sum == other.total_sum
        return False

print(nk.__version__)
# 3.15.1

```

Also, unrelated, but I'm curious about why you want to define a new 'spin' class. 
With the addition of constraints I'd hope this is no longer really needed anymore, and would like to know what are the use cases.

**Reply by alex-shilcusky** [reply]
Thank you checking this on your end and suggesting some possible solutions. I was checking my netket version incorrectly (VSCode terminal checking the wrong environment). With 3.15.1 nk.hilbert.constraint.DiscreteHilbertConstraint works fine. 

The use case is an idea of using a new spin basis for an old problem; it may simply need custom constraints or it may not work at all. 
Thanks again!
